node/hyperbahn/rate_limiting.js

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_QPS_LIMIT = 500;
+var DEFAULT_TOTAL_QPS_LIMIT = 1000;
+var DEFAULT_BUCKETS = 20;
+var TOTAL_QPS = '~~TotalQPS++';
+
+function RateLimiting(options) {
+    if (!(this instanceof RateLimiting)) {
+        return new RateLimiting(options);
+    }
+    var self = this;
+
+    assert(options.serviceDispatchHandler, 'serviceDispatchHandler is required for RateLimiting');
+    self.serviceDispatchHandler = options.serviceDispatchHandler;
+    self.egressNodes = options.serviceDispatchHandler.egressNodes;
+    self.timers = self.serviceDispatchHandler.channel.timers;
+
+    self.qpsLimits = options.qpsLimits || Object.create(null);
```

Comment:
I just did a search on Google. QPS seems be a well accepted concept for web services and bench marks. Actually, I am OK either way. I will make the change: QPS ==> RPS.